keps/sig-apps/4650-stateful-set-update-claim-template/README.md

```diff
@@ -0,0 +1,1087 @@
+  are updated to reflect the status of PVCs.
+
+With these changes, user can still use `kubectl rollout status` to monitor the update process,
+both for automated patching and for the PVCs that need manual intervention.
+
+### Updated Reconciliation Logic
+
+How to update PVCs:
+1. If `volumeClaimUpdatePolicy` is `InPlace`,
+   and if `volumeClaimTemplates` and actual PVC only differ in mutable fields
+   (`spec.resources.requests.storage`, `spec.volumeAttributesClassName`, `metadata.labels`, and `metadata.annotations` currently),
+   patch the PVC to the extent possible.
+   - `spec.resources.requests.storage` is patched to max(template spec, PVC status)
+     - Do not decreasing the storage size below its current status.
+       Note that decrease the size in PVC spec can help recover from a failed expansion if 
+       `RecoverVolumeExpansionFailure` feature gate is enabled.
+   - `spec.volumeAttributesClassName` is patched to the template value.
+   - `metadata.labels` and `metadata.annotations` are patched with server side apply.
+
+2. If it is not possible to make the PVC [compatible](#what-pvc-is-compatible),
+   do nothing. But when recreating a Pod and the corresponding PVC is deleting,
+   wait for the deletion then create a new PVC together with the new Pod (already implemented).
+<!--
+Tested on Kubernetes v1.28, and I can see this event:
+Warning  FailedCreate         3m58s (x7 over 3m58s)  statefulset-controller  create Pod test-rwop-0 in StatefulSet test-rwop failed error: pvc data-test-rwop-0 is being deleted
+-->
+
+3. Use either current or updated revision of the `volumeClaimTemplates` to create/update the PVC,
+   just like Pod template.
+
+When to update PVCs:
+1. before advancing `status.updatedReplicas` to the next replica,
+  check that the PVCs of the next replica are 
+   [compatible](#what-pvc-is-compatible) with the new `volumeClaimTemplates`.
+   If not, and if we are not going to patch it automatically,
+   wait for the user to delete/update the old PVC manually.
+
+2. When doing rolling update, A replica is considered ready if the Pod is ready
+   and all its volumes are not being updated in-place.
+   Wait for a replica to be ready for at least `minReadySeconds` before proceeding to the next replica.
+
+3. Whenever we check for Pod update, also check for PVCs update.
+   e.g.:
+   - If `spec.updateStrategy.type` is `RollingUpdate`,
+     update the PVCs in the order from the largest ordinal to the smallest.
+   - If `spec.updateStrategy.type` is `OnDelete`,
+     Only update the PVC when the Pod is deleted.
+   
+4. When patching the PVC, if we also re-create the Pod,
+   update the PVC after old Pod deleted, together with creating new pod.
+   Otherwise, if pod is not changed, update the PVC only.
+
+Failure cases: don't left too many PVCs being updated in-place. We expect to update the PVCs in order.
+
+- If the PVC update fails, we should block the update process.
```

Comment:
Which update? Can you be more specific, is it all of statefulset update or just pvc update?

Reply:
Should block the StatefulSet rollout process